File: app-play/app/assets/javascripts/app/directives/NewsDirective.js
(function() {
    var app = angular.module('shocktrade');

    /**
     * News Directive
     * @author lawrence.daniels@gmail.com
     * <news text="{{ myDate }}" /> ~> "5 days ago"
     */
    app.directive('news', ['$log', function($log) {
        return {
            restrict: 'E',
            scope: { class:'@class', content:'@content' },
            transclude: true,
            replace: false,
            template:
            '<div class="{{ class }}">' +
            '<a href="" ng-click="loadNewsQuote(\'{{symbol}}\')">' +
            '<nobr><span class="{{exchange}}">{{symbol}} {{change-arrow(change)}</span>' +
            '<span class="{{ changeClass(changePct) }}">{{ changePct }}</span></nobr>' +
            '</a>' +
            '</div>',
            link: function(scope, element, attrs) {
                scope.$watch( "content", function() {
                    var start = scope.content.indexOf("{$", start);
                    var end = scope.content.indexOf("$}", start);
                    if(start != -1 && end > start) {
                        var token = scope.content.substring(start+2, end - 1);
                        console.log("token = '" + token + "'");
                        var pcs = token.split("|");
                        scope.symbol = pcs[0];
                        scope.exchange = pcs[1];
                        scope.changePct = pcs[2];
                        console.log("symbol = " + scope.symbol + ", exchange = " + scope.exchange + ", changePct = " + scope.changePct);
                    }
                });
            }
        };
    }]);

    function replaceToken(encodedText) {
        var start = 0;
        do {
            start = encodedText.indexOf("${", start);
            var end = encodedText.indexOf("}$", end);
            if (start != -1 && end > start) {
                var token = encodedText.substring(start + 2, end - 1);
                console.log("token = '" + token + "'");
                var pcs = token.split("|");
                var symbol = pcs[0];
                var exchange = pcs[1];
                var changePct = pcs[2];
                console.log("symbol = " + symbol + ", exchange = " + exchange + ", changePct = " + changePct);
                encodedText = replace(encodedText, start, end + 2,
                    '<a href="" ng-click="loadNewsQuote(\'${symbol}\')">' +
                    '<nobr><span class="${exchange}">${symbol} ${change-arrow(change)}</span>' +
                    '<span class="${changeClass(changePct)}">${changePct}</span></nobr>' +
                    '</a>');
            }
            start = end;

        } while (start != -1 && start < encodedText.length);

        return encodedText;
    }

    function replace(source, start, end, replacement) {
        var s1 = source.substring(0, start);
        var s2 = source.substring(end, source.length);
        return s1 + replacement + s2;
    }

})();

File: app-angularjs/src/main/scala/com/shocktrade/javascript/directives/NewsDirective.scala
package com.shocktrade.javascript.directives

import org.scalajs.angularjs.Directive.{ElementRestriction, LinkSupport, TemplateSupport}
import org.scalajs.angularjs.sanitize.Sce
import org.scalajs.angularjs.{Attributes, Directive, JQLite, Scope}
import org.scalajs.dom.browser.console
import org.scalajs.nodejs.util.ScalaJsHelper._

import scala.scalajs.js
import scala.scalajs.js.annotation.ScalaJSDefined

/**
  * News Directive
  * @author lawrence.daniels@gmail.com
  * @example {{{ <news content="{{ myDate }}"></news> }}}
  */
class NewsDirective($sce: Sce) extends Directive with ElementRestriction with LinkSupport[NewsDirectiveScope] with TemplateSupport {
  override val scope = new NewsDirectiveInputs(`class` = "@class", content = "@content")
  override val template = """<div class="{{ class }}" ng-bind-html="output"></div>"""

  override def link(scope: NewsDirectiveScope, element: JQLite, attrs: Attributes): Unit = {
    scope.$watch("content", (newContent: js.UndefOr[String]) => scope.output = $sce.trustAsHtml(replaceTokens(newContent)))
  }

  private def replaceTokens(theContent: js.UndefOr[String]) = theContent.flat map { content =>
    val sb = new StringBuilder(content)
    var last = -1

    do {
      val start = sb.indexOf("{$", last)
      val end = sb.indexOf("$}", start)
      if (start != -1 && end > start) {
        last = end
        val token = sb.substring(start + 2, end - 1).trim
        token.split("[|]").map(_.trim).toList match {
          case symbol :: exchange :: changePct :: Nil =>
            console.log(s"symbol = $symbol, exchange = $exchange, changePct = $changePct")
            sb.replace(start, end + 2,
              s"""
              <a href="" ng-click="loadNewsQuote('$symbol')">
                <nobr>
                  <span class="$exchange">
                  $symbol <changearrow value="$changePct"></changearrow> {{ $changePct| quoteChange }}
                  </span>
                </nobr>
              </a>""")

          case values =>
            console.error(s"Invalid token values ('$token') => ${values.mkString(", ")}")
        }
      }
      else last = -1

    } while (last != -1)

    sb.toString()
  }

}

/**
  * News Directive Input Parameters
  * @author lawrence.daniels@gmail.com
  */
@ScalaJSDefined
class NewsDirectiveInputs(val `class`: String, val content: String) extends js.Object

/**
  * News Directive Scope
  * @author lawrence.daniels@gmail.com
  */
@js.native
trait NewsDirectiveScope extends NewsDirectiveInputs with Scope {
  // output fields
  var output: js.Any = js.native

}

